### PR TITLE
Laboratory work #1, Pavel Shkunov - 21FPL2

### PR DESCRIPTION
**Pull request by pa-shk**

### CODE REVIEW

File: lab_1_keywords_tfidf/main.py [hunk 1]
```diff
@@ -3,6 +3,23 @@
...
+import math
+
+
+def right_type(container: Union[list, tuple, dict, set], container_type: type,
+               elements_type: Union[type, dict[str, type]], allow_empty=True) -> bool:
+    """
+    Checks datatype of a container and its elements
+    """
+    empty_check = allow_empty or container
+    if type(container) == container_type and type(container) == dict:
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:16]: Использовать type - плохо

Comment (pa-shk) [reply lab_1_keywords_tfidf/main.py]: Исправлено.

**Review by artyomtugaryov — COMMENTED** [review]
Great job! Small fixes will make it awesome

File: lab_1_keywords_tfidf/main.py [hunk 2]
```diff
@@ -17,7 +49,10 @@ def clean_and_tokenize(text: str) -> Optional[list[str]]:
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if isinstance(text, str):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:52]: reverse the condition and return None immediately

File: lab_1_keywords_tfidf/main.py [hunk 3]
```diff
@@ -33,7 +68,9 @@ def remove_stop_words(tokens: list[str], stop_words: list[str]) -> Optional[list
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if right_type_container(tokens, list, str) and right_type_container(stop_words, list, str):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:71]: reverse the condition and return None immediately

File: lab_1_keywords_tfidf/main.py [hunk 4]
```diff
@@ -66,7 +105,12 @@ def get_top_n(frequencies: dict[str, Union[int, float]], top: int) -> Optional[l
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if (type_check(top, int)
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:108]: reverse the condition and return None immediately

File: lab_1_keywords_tfidf/main.py [hunk 5]
```diff
@@ -82,7 +126,11 @@ def calculate_tf(frequencies: dict[str, int]) -> Optional[dict[str, float]]:
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if (isinstance(frequencies, dict)
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:129]: reverse the condition and return None immediately

File: lab_1_keywords_tfidf/main.py [hunk 6]
```diff
@@ -99,7 +147,13 @@ def calculate_tfidf(term_freq: dict[str, float], idf: dict[str, float]) -> Optio
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if (right_dict(term_freq, keys_type=str, values_type=float, allow_empty=False)
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:150]: reverse the condition and return None immediately

File: lab_1_keywords_tfidf/main.py [hunk 7]
```diff
@@ -99,7 +147,13 @@ def calculate_tfidf(term_freq: dict[str, float], idf: dict[str, float]) -> Optio
...
+            and right_dict(idf, keys_type=str, values_type=float)):
+        tfidf_dict = {}
+        for term in term_freq:
+            tfidf_dict[term] = idf.get(term, math.log(47 / (0 + 1))) * term_freq[term]
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:154]: why `0 + 1` ?

**Review by artyomtugaryov — APPROVED** [review]
It is better to use CAPS for variables that defined outside of function (start.py)

**Pull request closed without merging**